presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergMetadata.java

```diff
@@ -598,23 +608,47 @@ public void rollback()
     @Override
     public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(ConnectorSession session, ConnectorTableHandle handle, Constraint constraint)
     {
-        IcebergTableHandle table = (IcebergTableHandle) handle;
+        IcebergTableHandle tableHandle = (IcebergTableHandle) handle;
+
         // TODO: Remove TupleDomain#simplify once Iceberg supports IN expression
         TupleDomain<IcebergColumnHandle> newDomain = constraint.getSummary()
                 .transform(IcebergColumnHandle.class::cast)
-                .simplify()
-                .intersect(table.getPredicate());
+                .intersect(tableHandle.getPredicate());
+
+        if (newDomain.isNone()) {
+            return Optional.empty();
+        }
 
-        if (newDomain.equals(table.getPredicate())) {
+        if (newDomain.equals(tableHandle.getPredicate())) {
             return Optional.empty();
         }
 
+        org.apache.iceberg.Table icebergTable = getIcebergTable(metastore, hdfsEnvironment, session, tableHandle.getSchemaTableName());
+
+        List<PartitionField> fields = icebergTable.spec().fields().stream()
+                .filter(field -> field.transform().isIdentity())
+                .collect(toImmutableList());
+
+        // Ensure partition specs in all manifests contain the identity fields from the predicate
+        if (!icebergTable.specs().values().stream().allMatch(spec -> spec.fields().containsAll(fields))) {
+            return Optional.empty();
+        }
+
+        Set<Integer> partitionSourceIds = icebergTable.spec().fields().stream()
+                .filter(field -> field.transform().isIdentity())
+                .map(PartitionField::sourceId)
+                .collect(toImmutableSet());
+        TupleDomain<ColumnHandle> remainingTupleDomain = newDomain.filter((column, domain) -> !partitionSourceIds.contains(column.getId()))
```

Comment:
Maybe do this as
```java
BiPredicate<IcebergColumnHandle, Domain> contains = (column, domain) -> partitionSourceIds.contains(column.getId());
TupleDomain<ColumnHandle> remainingTupleDomain = newDomain.filter(contains.negate());
TupleDomain<IcebergColumnHandle> enforcedTupleDomain = newDomain.filter(contains);

Reply:
Nice!  Changed.  The `remainingTupleDomain` needs suffix `.transform(ColumnHandle.class::cast)` in this formulation.